packages/plugins/typescript/effect/src/visitor.ts

```diff
@@ -0,0 +1,198 @@
+import autoBind from 'auto-bind';
+import { type GraphQLSchema, type OperationDefinitionNode, print } from 'graphql';
+import {
+  type ClientSideBasePluginConfig,
+  ClientSideBaseVisitor,
+  DocumentMode,
+  LoadedFragment,
+  type RawClientSideBasePluginConfig,
+} from '@graphql-codegen/visitor-plugin-common';
+
+export interface EffectPluginConfig extends ClientSideBasePluginConfig {}
+
+const additionalStaticContent = (documentMode: DocumentMode) => `
+export type GraphQLSuccessResponse<A> = {
+  body: ExecutionResult & { data: A };
+  headers: Record<string, string>;
+};
+
+export type GraphQLErrorResponse = {
+  body: Omit<ExecutionResult, 'data'>;
+  headers: Record<string, string>;
+};
+
+export class MissingDataGraphQLResponseError extends Data.TaggedError(
+  'MissingDataGraphQLResponseError',
+)<GraphQLErrorResponse> {}
+
+export type GraphQLOperationOptions = {
+  preferredOpName?: string;
+};
+
+type GraphQLOperationArgs = {
+  document: ${documentMode === DocumentMode.string ? 'string' : 'DocumentNode'};
+  fallbackOperationName: string;
+};
+
+// https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#legacy-watershed
+const Accept = 'application/graphql-response+json; charset=utf-8, application/json; charset=utf-8';
+
+const makeGraphQLClient = Http.client.mapRequest(
+  Http.request.setHeaders({ Accept, 'Content-Type': 'application/json' }),
+);
+
+const makeGraphQLOperation =
+  <Vars, Data>({ document, fallbackOperationName }: GraphQLOperationArgs) =>
+  (variables: Vars, opts?: GraphQLOperationOptions) => {
+    const operationName = opts?.preferredOpName ?? fallbackOperationName;
+    const query = ${documentMode === DocumentMode.string ? 'document' : 'print(document)'};
+
+    return Effect.flatMap(Http.client.Client, client =>
+      Http.request.post('').pipe(
+        Http.request.jsonBody({
+          query,
+          operationName,
+          variables,
+        }),
+        Effect.flatMap(Http.client.filterStatusOk(makeGraphQLClient(client))),
+        Effect.flatMap(
+          Http.response.schemaJson(
```

Comment:
You could move this the top level so the schema isn't rebuilt for each request.

Reply:
Hmm, why not just use `Effect.flatMap(res => Effect.map(res.json, body => ({ headers: res.headers, body }))` here? Don't really need a schema.

Reply:
You could move some bits to the client constructor:

```ts
import * as Http from "@effect/platform/HttpClient"
import { Effect } from "effect"

const headers = {
  Accept:
    "application/graphql-response+json; charset=utf-8, application/json; charset=utf-8",
  "Content-Type": "application/json",
}

const makeGraphQLClient = <R, E>(
  client: Http.client.Client.WithResponse<R, E>,
) =>
  Http.client.mapRequest(client, Http.request.setHeaders(headers)).pipe(
    Http.client.filterStatusOk,
    Http.client.mapEffectScoped((res) =>
      res.json.pipe(
        Effect.map((body) => ({
          headers: res.headers,
          body: body as any,
        })),
        Effect.flatMap(res => {
          const body = res.body as ExecutionResult<Data>;
          const { headers } = res;
          if (body.data) return Effect.succeed(res as GraphQLSuccessResponse<Data>);
          return Effect.fail(new MissingDataGraphQLResponseError({ body, headers }));
        }),
      ),
    ),
  )
```